DOC: update vendored numpydoc version

jorisvandenbossche opened the pull request:
The first commit only updates our vendored version with the upstream one. 

Additional commits make it work with our docs:

- edit numpydoc to allow member listing for attributes (this is a change (and the only) to the source code of numpydoc that is not upstream: discussed in https://github.com/numpy/numpydoc/pull/106)
- numpydoc settings: use `numpydoc_use_blockquotes=False` for now for compatibility (it is deprecated). To fix this, we need to fix: css styling of description lists + warnings generated due to `**kwargs`

It is probably worth to look at the separate commits (and would also like to keep them when merging)

jorisvandenbossche committed:
DOC: update vendored numpydoc version

Numpydoc upstream at 21a194e9b42ebe95a70475e35cb69bcb4801ff36

jorisvandenbossche committed:
DOC: use numpydoc_use_blockquotes=False for compatibility

pep8speaks commented:
Hello @jorisvandenbossche! Thanks for submitting the PR.





 - In the file [`doc/sphinxext/numpydoc/docscrape.py`](https://github.com/pandas-dev/pandas/blob/72e21069d9035a78a425dec3772b00b184892f0d/doc/sphinxext/numpydoc/docscrape.py), following are the PEP8 issues :

> [Line 567:21](https://github.com/pandas-dev/pandas/blob/72e21069d9035a78a425dec3772b00b184892f0d/doc/sphinxext/numpydoc/docscrape.py#L567): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent

 - In the file [`doc/sphinxext/numpydoc/docscrape_sphinx.py`](https://github.com/pandas-dev/pandas/blob/72e21069d9035a78a425dec3772b00b184892f0d/doc/sphinxext/numpydoc/docscrape_sphinx.py), following are the PEP8 issues :

> [Line 38:80](https://github.com/pandas-dev/pandas/blob/72e21069d9035a78a425dec3772b00b184892f0d/doc/sphinxext/numpydoc/docscrape_sphinx.py#L38): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 354:25](https://github.com/pandas-dev/pandas/blob/72e21069d9035a78a425dec3772b00b184892f0d/doc/sphinxext/numpydoc/docscrape_sphinx.py#L354): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ':'



 - In the file [`doc/sphinxext/numpydoc/tests/test_docscrape.py`](https://github.com/pandas-dev/pandas/blob/72e21069d9035a78a425dec3772b00b184892f0d/doc/sphinxext/numpydoc/tests/test_docscrape.py), following are the PEP8 issues :

> [Line 458:1](https://github.com/pandas-dev/pandas/blob/72e21069d9035a78a425dec3772b00b184892f0d/doc/sphinxext/numpydoc/tests/test_docscrape.py#L458): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 595:1](https://github.com/pandas-dev/pandas/blob/72e21069d9035a78a425dec3772b00b184892f0d/doc/sphinxext/numpydoc/tests/test_docscrape.py#L595): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 731:5](https://github.com/pandas-dev/pandas/blob/72e21069d9035a78a425dec3772b00b184892f0d/doc/sphinxext/numpydoc/tests/test_docscrape.py#L731): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 741:80](https://github.com/pandas-dev/pandas/blob/72e21069d9035a78a425dec3772b00b184892f0d/doc/sphinxext/numpydoc/tests/test_docscrape.py#L741): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (124 > 79 characters)
> [Line 1184:80](https://github.com/pandas-dev/pandas/blob/72e21069d9035a78a425dec3772b00b184892f0d/doc/sphinxext/numpydoc/tests/test_docscrape.py#L1184): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 79 characters)
> [Line 1186:5](https://github.com/pandas-dev/pandas/blob/72e21069d9035a78a425dec3772b00b184892f0d/doc/sphinxext/numpydoc/tests/test_docscrape.py#L1186): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent

jorisvandenbossche committed:
DOC: change to numpydoc: add option to use member listing for attributes

TomAugspurger commented on doc/sphinxext/numpydoc/docscrape_sphinx.py:
```diff
@@ -33,6 +33,7 @@ def load_config(self, config):
         self.use_plots = config.get('use_plots', False)
         self.use_blockquotes = config.get('use_blockquotes', False)
         self.class_members_toctree = config.get('class_members_toctree', True)
+        self.attributes_as_param_list = config.get('attributes_as_param_list', True)
```
Will this ensure that Attributes is formatted like a table?

jorisvandenbossche replied:
Yes (updated the top post). For now, that is the only change I made to the numpydoc source (it's in a separate commit).

But, if I am the only one that finds that better, I am OK to leave that.

jorisvandenbossche replied:
Well, there isn't ;-) That's something I added here (but with only a few lines: https://github.com/pandas-dev/pandas/pull/20383/commits/075438f4b5ba14df925e79f00854c2d4b61e0cae), and that's what I proposed to do in https://github.com/numpy/numpydoc/pull/106, so it would be nice to have this in upstream)

jorisvandenbossche commented:
We still have the problem we fixed with this hack: https://github.com/pandas-dev/pandas/commit/70c9d318ed4ef32e1c294a4dbefcb783cda19dba, related to the property's that are None.

However, due to recent work, I think only `.index` and `.columns` are left for this (there are others there as well that don't show up properly, but that is because of the `if pydoc.getdoc(param_obj)` check and they have no docstring .. so that is easily fixable).

jacko8944 commented:
Lundy06032013@gmail.com

jorisvandenbossche commented:
`.index` and `.columns` should be fixed with https://github.com/pandas-dev/pandas/pull/20385

jorisvandenbossche commented:
Updated this. 

The remaining related warnings I think are:

/tmp/doc/source/generated/pandas.Index.holds_integer.rst: WARNING: document isn't included in any toctree
/tmp/doc/source/generated/pandas.Index.is_type_compatible.rst: WARNING: document isn't included in any toctree
/tmp/doc/source/generated/pandas.Series.imag.rst: WARNING: document isn't included in any toctree
/tmp/doc/source/generated/pandas.Series.real.rst: WARNING: document isn't included in any toctree

this seems to be because they are not listed manually in api.rst, and are also not linked in the class docstring page (because they have no docstring they appear as plain table entry, not link to the docstring page. With as a result there is nowhere a link to that page).

TomAugspurger commented:
Would adding them to the `hidden` toctree at the bottom of `api.rst` be OK? It'd fix the warnings. Not sure what happens to the links...

jorisvandenbossche commented:
> Would adding them to the hidden toctree at the bottom of api.rst be OK? It'd fix the warnings. Not sure what happens to the links...

So now I need to remember again why we had those entries there at the bottom of api.rst ... Which you added in https://github.com/pandas-dev/pandas/pull/18202. Because this are attributes that were not yet included in api.rst before, and it is only now because they were no longer included in their class docstring page, there was a warning. 
But actually just giving them a proper docstring, should put them back in their class docstring page, and so fix the warning. So that is the more proper fix I think.

jorisvandenbossche committed:
DOC: change to numpydoc: use isdatadescriptor instead of isgetsetdescriptor to check attributes

This is needed to ensure .columns / .index / cached properties are recognized as an attribute

jorisvandenbossche commented:
OK, so it seems that as long as they are properly included in the class docstring page toctree, that is also fine, so in principle then we can get rid of the full list of hidden items in api.rst (but for that, we need to add docstrings to some of those methods). That's the proper fix for the warnings, but will leave that for another PR (now added them to the hidden list in api.rst)

jorisvandenbossche commented:
Travis doc build looks good
